File: src/percolation/PercolationStats.java
import java.lang.IllegalArgumentException;
import java.util.Random;

public class PercolationStats {
    private int N;
    private int T;

    private double[] measurements;

    private double mean;
    private double stddev;
    private double confidenceLo;
    private double confidenceHi;

    public PercolationStats(int N, int T) {  // perform T independent experiments on an N-by-N grid
        if (N <= 0 || T <= 0)
            throw new IllegalArgumentException("N and T must be larger than 0");

        this.N = N;
        this.T = T;

        measurements = new double[T];

        action();
        calcStats();
    }

    private void action() {
        for (int i = 0; i < T; i++) {
            measurements[i] = run() * 1.0/(N*N);
        }
    }

    private int run() {
        Percolation p = new Percolation(N);

        int count = 0;
        int[] sites = new int[N*N];
        for (int i = 0; i < sites.length; i++)
            sites[i] = i;

        int available = sites.length;

        Random r = new Random();
        while (!p.percolates()) {

            //get site
            int n = r.nextInt(available);
            int site = sites[n];
            int i = site/N;
            int j = site % N;

            //open it
            p.open(i + 1, j + 1);

            //swap to tne end
            int temp = sites[n];
            sites[n] = sites[available-1];
            sites[available-1] = temp;

            available--;

            count++;
        }

        return count;
    }

    private void calcStats() {
        double sum = 0;
        for (int i = 0; i < T; i++)
            sum += measurements[i];

        mean =  sum/T;

        double delta = 0;
        for (int i = 0; i < T; i++)
            delta += (measurements[i] - mean) * (measurements[i] - mean);

        stddev = Math.sqrt(delta/(T-1));

        confidenceLo = mean - 1.96 * stddev/Math.sqrt(T);

        confidenceHi = mean + 1.96 * stddev/Math.sqrt(T);

    }

    public double mean() {                  // sample mean of percolation threshold
       return mean;
    }
    public double stddev() {             // sample standard deviation of percolation threshold
       return stddev;
    }

    public double confidenceLo() {           // low  endpoint of 95% confidence interval
        return confidenceLo;
    }

    public double confidenceHi() {          // high endpoint of 95% confidence interval
        return confidenceHi;
    }

    public static void main(String[] args) {  // test client (described below)
        if (args.length != 2)
            throw new IllegalArgumentException("args must be 2");

        int N = Integer.parseInt(args[0]);
        int T = Integer.parseInt(args[1]);

        PercolationStats ps = new PercolationStats(N, T);

        System.out.printf("mean                    = %f\n", ps.mean());
        System.out.printf("stddev                  = %f\n", ps.stddev());
        System.out.printf("95%% confidence interval = %f, %f\n" , ps.confidenceLo(), ps.confidenceHi());
    }
}

File: src/percolation/percolation_stats.rs
extern crate rand;

//use self::rand::Rng;

use super::percolation::Percolation;

pub struct PercolationStats {
    n : usize,
    t : usize,
    measurements: Vec<f64>,

    mean: f64,
    stddev: f64,
    confidence_lo: f64,
    confidence_hi: f64
}

impl PercolationStats {
    pub fn new(n:usize, t: usize) -> PercolationStats {
        assert!(n > 1 && t > 1);

        let mut ps = PercolationStats {n: n,
                          t: t,
                          measurements:vec![0.0; t],
                          mean: 0.0,
                          stddev: 0.0,
                          confidence_lo: 0.0,
                          confidence_hi: 0.0};

       ps.performe_runs();
       ps.calc_stats();
       ps
    }

    fn performe_runs(&mut self) {
        let size = (self.n * self.n) as f64;
        for i in 0..self.t {
            self.measurements[i] = self.run() / size;
        }
    }

    fn run(&self) -> f64 {
        let mut count = 0.0;
        let mut p = Percolation::new(self.n);
        let mut sites = (0..(self.n * self.n)).collect::<Vec<_>>();

        let mut available_size = sites.len();

        //let mut rng = rand::thread_rng();
        //et mut rng = rand::thread_rng();
        while !p.percolates() {

            let index = rand::random::<usize>() % available_size; //rng.get_range(0, available_size);

            //get site
            let site = sites[index];

            let i = site / self.n;
            let j = site % self.n;

            //open site
            p.open(i + 1, j + 1);

            //move the opened site to end
            sites.swap(index, available_size - 1);

            available_size -= 1;
            count += 1.0;
        }

        count
    }

    fn calc_stats(&mut self) {
        let t = self.t as f64;
        let t1 = (self.t - 1) as f64;

        let sum: f64 = self.measurements.iter().fold(0.0, |acc, x| acc + x);
        let mean = sum /(self.t as f64);

        let delta: f64 = self.measurements.iter()
                            .fold(0.0, |acc, x| acc + (x - mean) * (x - mean));

        let stddev = (delta / t1).sqrt();

        self.mean = mean;
        self.stddev = stddev;

        let sqrt_t = t.sqrt();

        self.confidence_lo = mean - 1.96 * stddev / sqrt_t;
        self.confidence_hi = mean + 1.96 * stddev / sqrt_t
    }

    pub fn mean(&self) -> f64 {
        self.mean
    }

    pub fn stddev(&self) -> f64 {
        self.stddev
    }

    pub fn confidence_lo(&self) -> f64 {
        self.confidence_lo
    }

    pub fn confidence_hi(&self) -> f64 {
        self.confidence_hi
    }
}
